### PR TITLE
feat: add gradient accumulation support

### PR DESCRIPTION
**Pull request by irisTa56**
## Summary / Motivation

Adds gradient accumulation support to the training script.
This allows simulating larger batch sizes without increasing GPU memory usage, which is useful when training large models or when memory is limited.

## What changed

- Added `gradient_accumulation_steps` configuration parameter to `TrainPipelineConfig` (default: 1)
- Updated `update_policy()` to use `accelerator.accumulate()` context manager
- Updated effective batch size calculation in logging and `MetricsTracker`
- Training loop now skips evaluation and checkpointing during gradient accumulation steps (i.e., treats `step` as an optimizer step)
- Added validation that rejects CUDAGraphs compile modes (`max-autotune`, `reduce-overhead`) when `gradient_accumulation_steps > 1`
- Added gradient accumulation section to multi-GPU training docs

## How was this tested (or how to run locally)

Reproduce with a quick example or CLI (if applicable):

```bash
$ lerobot-train \
  --dataset.repo_id=lerobot/svla_so101_pickplace \
  --policy.type=act \
  --policy.repo_id=example/my_policy \
  --policy.push_to_hub=false \
  --batch_size=8 \
  --gradient_accumulation_steps=4 \
  --log_freq=10 \
  --steps=50

# with compilation
$ lerobot-train \
  --dataset.repo_id=lerobot/svla_so101_pickplace \
  --policy.type=smolvla \
  --policy.repo_id=example/my_policy \
  --policy.pretrained_path=lerobot/smolvla_base \
  --policy.compile_model=true \
  --policy.compile_mode=max-autotune-no-cudagraphs \
  --policy.push_to_hub=false \
  --batch_size=8 \
  --gradient_accumulation_steps=4 \
  --log_freq=10 \
  --steps=50
```

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [ ] All tests pass locally (`pytest`)
  - Some tests were skipped in my local environment: `1586 passed, 123 skipped`
- [x] Documentation updated
- [ ] CI is green

### CODE REVIEW

File: src/lerobot/scripts/lerobot_train.py [hunk 1]
```diff
@@ -87,38 +84,42 @@ def update_policy(
...
+        optimizer.zero_grad()
 
-    optimizer.zero_grad()
-
-    # Step through pytorch scheduler at every batch instead of epoch
-    if lr_scheduler is not None:
-        lr_scheduler.step()
+        # Step through pytorch scheduler at every batch instead of epoch
+        if lr_scheduler is not None:
+            lr_scheduler.step()  # Accelerator automatically skips during accumulation
```
[the remark this answers is not in the document]
Comment (irisTa56) [reply src/lerobot/scripts/lerobot_train.py]: fixed https://github.com/huggingface/lerobot/pull/2646/commits/b727b8f40e2cadaca0e72951aed1a3b4e05c1893

File: tests/training/test_update_policy.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+
+def test_update_policy_gradient_accumulation_equivalence():
+    gradient_accumulation_steps = 4
+    batch_size = 4
+
+    policy_no_accum, optimizer_no_accum, lr_scheduler_no_accum = create_policy_and_optimizer()
+    policy_with_accum, optimizer_with_accum, lr_scheduler_with_accum = create_policy_and_optimizer()
+
+    batches_small = [{"observation": torch.randn(batch_size, 10)} for _ in range(gradient_accumulation_steps)]
```
[the remark this answers is not in the document]
Comment (irisTa56) [reply tests/training/test_update_policy.py]: fixed https://github.com/huggingface/lerobot/pull/2646/commits/06b047b0bc17bd17352a07a09cd9c84ebe928c53

**Comment by jadechoghari** [comment]
Hey @irisTa56 that's a very useful feature! I did it locally to train bigger models. Could you maybe resolve the conflicts with main, so we kickoff a review?
Thanks!

**Comment by irisTa56** [comment]
@jadechoghari 
Thanks for the feedback! I've resolved the merge conflicts with main and also added code comments for the newly introduced RA-BC feature. The PR should be ready for review now.

**Comment by jadechoghari** [comment]
Thanks! @irisTa56, the tests in this PR seem unnecessary since we’re using the accelerate api, and testing this functionality is the responsibility of accelerate itself. Removing the accelerate-related tests would make the PR cleaner and shorter 😄

**Comment by irisTa56** [comment]
> Thanks! @irisTa56, the tests in this PR seem unnecessary since we’re using the accelerate api, and testing this functionality is the responsibility of accelerate itself. Removing the accelerate-related tests would make the PR cleaner and shorter 😄

Thanks for the feedback! I agree and have removed the tests.

File: tests/utils/test_logging_utils.py [hunk 3]
```diff
@@ -82,6 +83,30 @@ def test_metrics_tracker_step(mock_metrics):
...
+    assert tracker.samples == 0
+
+    tracker.step()
+    expected_samples = 32 * accelerator.num_processes * gradient_accumulation_steps
+    assert tracker.steps == 1
+    assert tracker.samples == expected_samples
+    assert tracker.episodes == expected_samples / (1000 / 50)
+    assert tracker.epochs == expected_samples / 1000
+
+
```
Comment (jadechoghari) [inline tests/utils/test_logging_utils.py:123]: same here, I don’t think this test is necessary, we’re already testing the metrics stepper itself in other tests, and this ends up indirectly testing accelerate’s behavior rather than our own logic

Comment (irisTa56) [reply tests/utils/test_logging_utils.py:123]: Makes sense. Existing tests already cover the metrics stepper logic itself. Removed the extra tests https://github.com/huggingface/lerobot/pull/2646/commits/d365eb8d1ec828adc6c9c75495d585ec8a513de3

**Comment by jadechoghari** [comment]
is the pr ready @irisTa56

**Comment by irisTa56** [comment]
@jadechoghari 
Yes, it’s ready for review.

**Comment by sdfbiasdf** [comment]
Hi @irisTa56 , great work on this PR! I've been testing it with pi05 + compile_model=true and found a  CUDAGraphs compatibility issue.

 Problem: When gradient_accumulation_steps > 1 and compile_model=true, the default compile_mode="max-autotune" uses CUDAGraphs, which crashes because multiple forward passes before backward overwrite tensors captured in the graph.

 Fix: I made compile_mode default to None across all policy configs, with runtime resolution:
 - No accumulation → each policy's modeling file resolves to its original default (max-autotune, reduce-overhead, etc.)
 - With accumulation → lerobot_train.py auto-resolves to max-autotune-no-cudagraphs
 - If user explicitly sets a CUDAGraphs mode + accumulation → ValueError with clear message

 Changes: sdfbiasdf/lerobot@feat/grad-accum

 Includes:
 - compile_mode fix for all 5 policies (pi0, pi05, pi0_fast, smolvla, diffusion)
 - Tests (14 cases, all passing)
 - Documentation for gradient accumulation in training guide

  Feel free to cherry-pick if useful!

**Comment by sdfbiasdf** [comment]
Hi @irisTa56 — quick follow-up to my earlier comment. I refactored the fix into a cleaner, more extensible shape on the same branch. Two main changes:

**1. Per-policy resolution via the config base class.** Instead of a uniform "always fall back to `max-autotune-no-cudagraphs`" rule in `lerobot_train.py`, each policy now declares its own `DEFAULT_COMPILE_MODE` and `SAFE_COMPILE_MODE` as `ClassVar`s, and `PreTrainedConfig.resolve_compile_mode()` is the single source of truth. The train script becomes policy-agnostic and adding a new policy is just two class attributes.

**2. Diffusion gets `"default"` instead of `"max-autotune-no-cudagraphs"`.** This was the bug in my first version: `reduce-overhead` is the documented diffusion default precisely because the U-Net runs small repetitive loops where `max-autotune`-style tuning is overkill. Forcing `max-autotune-no-cudagraphs` when accumulation is on would have run an expensive autotune search against that intent. `"default"` is the lightweight CUDAGraphs-free analogue and preserves the original spirit.

| Policy | DEFAULT | SAFE (with grad_accum > 1) |
|---|---|---|
| pi0 / pi05 / pi0_fast / smolvla | `max-autotune` | `max-autotune-no-cudagraphs` |
| diffusion | `reduce-overhead` | `default` |

Bonus: tests now call `config.resolve_compile_mode()` directly (no helper duplication), parametrized over all 5 policies × both CUDAGraphs modes.

Same offer as before — just wanted to flag the diffusion case specifically.

Refactor commit: https://github.com/sdfbiasdf/lerobot/commit/3b771c23
Docs commit: https://github.com/sdfbiasdf/lerobot/commit/f1e349fd

**Comment by irisTa56** [comment]
Good catch, @sdfbiasdf
Thanks for digging into the compile mode interaction and the per-policy resolution!

The automatic per-policy resolution you built is solid work.
Since it touches multiple config files, I think it'd be easier to review as a separate PR.
Would you be open to submitting that as a follow-up?

For this PR, to keep it focused on the core gradient accumulation feature, I added a hard error when CUDAGraphs compile modes are combined with `gradient_accumulation_steps > 1`, along with documentation.

Thanks again!

**Comment by sdfbiasdf** [comment]
Sure, happy to open it as a follow-up. Thanks for the credit!

**Comment by sdfbiasdf** [comment]
Hi @jadechoghari — any chance of a review here? I have a small follow-up ready to open once this lands.

**Comment by irisTa56** [comment]
Closing this in favor of #3980, which supersedes it and carries the gradient accumulation implementation forward plus the per-policy compile mode resolution follow-up we discussed here.
